### PR TITLE
vcpkg ci: Parse ci.baseline.txt to determine regressions

### PR DESCRIPTION
**Pull request by autoantwort**
Probably you don't want to use iostreams. Should I add `write_unlocalized_text_to_stderr` to write to stderr?   

This makes afaik the whole powershell parsing stuff in the ci obsolete

### CODE REVIEW

**Comment by BillyONeal** [comment]
Stop doing things on my todo list before I do them 😅

Thanks for the submission! Will take a look soon

File: src/vcpkg/commands.ci.cpp [hunk 1]
```diff
@@ -678,6 +780,38 @@ namespace vcpkg::Commands::CI
...
+                for (auto&& port_result : result.summary.results)
+                {
+                    switch (port_result.build_result.code)
+                    {
+                        case Build::BuildResult::BUILD_FAILED:
+                        case Build::BuildResult::POST_BUILD_CHECKS_FAILED:
+                        case Build::BuildResult::FILE_CONFLICTS:
+                            if (expected_failures.find(port_result.spec) == expected_failures.end())
+                            {
+                                std::cerr
```
Comment (BillyONeal) [inline src/vcpkg/commands.ci.cpp:796]: I think this should also use print2, and we certainly should avoiding new use of iostreams if at all possible.

**Comment by BillyONeal** [comment]
Sorry it has taken me so long to get to this; I didn't leave comments back at the time because I thought I was going to respond quickly with the bits I wanted fixed but that didn't work out.

It looks like Nicole already commented about effectively everything I was going to say, except I would like to see a unit test put around `parse_ci_baseline`. At that point I would be happy to merge this and make the corresponding infrastructure changes.

If don't want to do that because you've been kept hanging too long I understand :)

**Comment by BillyONeal** [comment]
> If don't want to do that because you've been kept hanging too long I understand :)

Are you interested in that? Otherwise I can push that into this branch.

**Comment by autoantwort** [comment]
> Otherwise I can push that into this branch.

That would be cool. I am currently writing exams and don't have that much time :)

**Comment by BillyONeal** [comment]
OK I have to fix the bug where the "one liner" doesn't update correctly and this is next on my list.

File: src/vcpkg/commands.ci.cpp [hunk 2]
```diff
@@ -74,6 +77,53 @@ namespace
...
+            CiBaselineLine parsed_line;
+
+            auto colon_loc = line.find(':');
+            Checks::check_exit(VCPKG_LINE_INFO, colon_loc != std::string::npos, "Line '%s' must contain a ':'", line);
+            parsed_line.port_name = Strings::trim(StringView{line.data(), line.data() + colon_loc}).to_string();
+
+            auto equal_loc = line.find('=', colon_loc + 1);
+            Checks::check_exit(VCPKG_LINE_INFO, equal_loc != std::string::npos, "Line '%s' must contain a '='", line);
+            parsed_line.triplet_name =
+                Strings::trim(StringView{line.data() + colon_loc + 1, line.data() + equal_loc}).to_string();
```
Comment (BillyONeal) [inline src/vcpkg/commands.ci.cpp:109]: FYI I'm changing this behavior because it fails, for example, with

```
catch-classic:arm64-windows      = skip
catch-classic:arm-uwp            = skip
catch-classic:x64-linux          = skip
catch-classic:x64-osx            = skip
catch-classic:x64-uwp            = skip
catch-classic:x64-windows        = skip
catch-classic:x64-windows-static = skip
catch-classic:x64-windows-static-md=skip
catch-classic:x86-windows        = skipcatch-classic:arm64-windows      = skip
catch-classic:arm-uwp            = skip
catch-classic:x64-linux          = skip
catch-classic:x64-osx            = skip
catch-classic:x64-uwp            = skip
catch-classic:x64-windows        = skip
catch-classic:x64-windows-static = skip
catch-classic:x64-windows-static-md=skip
catch-classic:x86-windows        = skip
```

**Review by BillyONeal — APPROVED** [review]

**Review by BillyONeal — CHANGES_REQUESTED** [review]
Whoops, I forgot to get rid of cerr

**Review by BillyONeal — APPROVED** [review]

File: include/vcpkg/base/sortedvector.h [hunk 3]
```diff
@@ -43,9 +58,53 @@ namespace vcpkg
...
+            if (!std::is_sorted(m_data.begin(), m_data.end(), m_comp))
+            {
+                std::sort(m_data.begin(), m_data.end(), m_comp);
+            }
+
+            uniqueify();
+        }
+
+        std::vector<Ty> m_data;
+        Compare m_comp;
```
[the remark this answers is not in the document]
Comment (strega-nil) [reply include/vcpkg/base/sortedvector.h:108]: Also, WHY THE HECK DOESN'T MSVC SUPPORT `[[no_unique_address]]`?????

File: locales/messages.json [hunk 4]
```diff
@@ -47,6 +47,12 @@
   "_ChecksLineInfo.comment": "{Locked}",
   "ChecksUnreachableCode": "unreachable code was reached",
   "ChecksUpdateVcpkg": "updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.",
+  "CiBaselineRegression": "REGRESSION: {spec} failed with {build_result}. If expected, add {spec}=fail to {path}.",
+  "_CiBaselineRegression.comment": "example of {spec} is 'zlib:x64-windows'.\nexample of {build_result} is 'One of the BuildResultXxx messages (such as BuildResultSucceeded/SUCCEEDED)'.\nexample of {path} is '/foo/bar'.\n",
```
Comment (strega-nil) [inline locales/messages.json:51]: this should be `{Locked}`.

File: src/vcpkg/binarycaching.cpp [hunk 5]
```diff
@@ -92,7 +90,7 @@ namespace
             std::string segment;
             for (;;)
             {
-                auto n = match_until([](char32_t ch) { return ch == ',' || ch == '`' || ch == ';'; });
+                auto n = match_while([](char32_t ch) { return ch != ',' && ch != '`' && ch != ';'; });
```
Comment (strega-nil) [inline src/vcpkg/binarycaching.cpp:93]: why this change?

File: include/vcpkg/base/parse.h [hunk 6]
```diff
@@ -81,6 +80,7 @@ namespace vcpkg::Parse
         static constexpr bool is_icase_alpha(char32_t ch) { return is_lower_alpha(ch) || is_upper_alpha(ch); }
         static constexpr bool is_ascii_digit(char32_t ch) { return ch >= '0' && ch <= '9'; }
         static constexpr bool is_lineend(char32_t ch) { return ch == '\r' || ch == '\n' || ch == Unicode::end_of_file; }
+        static constexpr bool is_not_lineend(char32_t ch) { return !is_lineend(ch); }
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply include/vcpkg/base/parse.h:83]: This change eliminates blah_until because this was the only caller.

File: locales/messages.json (hunk 4 above)
Comment (BillyONeal) [reply locales/messages.json:51]: I wasn't sure if it should be. The "spec=fail" part needs to be but there's no reason the rest of the message couldn't be localized.

File: src/vcpkg/binarycaching.cpp (hunk 5 above)
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:93]: To merge match_until and match_zero_or_more into one function.

**Comment by BillyONeal** [comment]
I sucked out https://github.com/microsoft/vcpkg-tool/pull/447/ "Delete Parse::" to make this easier to review if you want to land that one first.

File: include/vcpkg/base/sortedvector.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply include/vcpkg/base/sortedvector.h:108]:
```
C:\Dev\vcpkg-tool\include\vcpkg\base\sortedvector.h(108): warning C4848: support for attribute 'no_unique_address' in C++17 and earlier is a vendor extension
  C:\Dev\vcpkg-tool\include\vcpkg\base\sortedvector.h(109): note: see reference to class template instantiation 'vcpkg::SortedVector<Ty,Compare>' being compiled
```

My fears were right :(

File: include/vcpkg/base/parse.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (strega-nil) [reply include/vcpkg/base/parse.h:76]: I've gotta be honest, I much prefer having `blah_until`; I think it's much more readable.

File: include/vcpkg/base/sortedvector.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (strega-nil) [reply include/vcpkg/base/sortedvector.h:108]: Can we inherit from `Compare` then?

File: locales/messages.json (hunk 4 above)
Comment (strega-nil) [reply locales/messages.json:51]: mmh, okay

File: include/vcpkg/base/parse.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply include/vcpkg/base/parse.h:76]: OK, I'll revert that part of the change but keep the _while rename part. Stand by...

**Review by strega-nil — APPROVED** [review]
Approved, but I still prefer to have `match_until` as well.

**Comment by BillyONeal** [comment]
>Can we inherit from Compare then?

I really don't want to do that because it gives SortedVector an `operator()`. I can see an argument for a compressed_pair optimization in the future in order to implement this but I think that is a separate improvement and we shouldn't block over that.

Alternately, we can change the SortedVector design such that the caller always needs to pass the comparison functor every time they do something that needs it.... let me try that.

File: src/vcpkg/binarycaching.cpp (hunk 5 above)
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp]: Reverted.

File: include/vcpkg/base/parse.h [hunk 7]
```diff
@@ -137,10 +133,10 @@ namespace vcpkg
         void add_warning(LocalizedString&& message, const SourceLoc& loc);
 
         const ParseError* get_error() const { return m_messages.error.get(); }
-        std::unique_ptr<ParseError> extract_error() { return std::move(m_messages.error); }
+        std::unique_ptr<ParseError> extract_error() && { return std::move(m_messages.error); }
```
Comment (strega-nil) [inline include/vcpkg/base/parse.h:136]: These are intentionally non-rvalue - `extract_blah` in vcpkg means "do a move inside" (see also `LocalizedString::extract_data()`)

File: include/vcpkg/base/sortedvector.h [hunk 8]
```diff
@@ -1,36 +1,56 @@
...
+#include <functional>
+#include <initializer_list>
+#include <iterator>
 #include <vector>
 
 // Add more forwarding functions to the m_data std::vector as needed.
 namespace vcpkg
 {
-    template<class T>
+    template<class Ty>
```
Comment (strega-nil) [inline include/vcpkg/base/sortedvector.h:12]: I'm sorry, I'd prefer the old version that just takes a `Compare`; I care more about the API here than the extra eight bytes.

File: src/vcpkg/base/parse.cpp [hunk 9]
```diff
@@ -105,6 +127,68 @@ namespace vcpkg
...
+            }
+
+            if (first1 == last1 || *first1 != static_cast<char32_t>(*first2))
+            {
+                return false;
+            }
+
+            ++first1;
+            ++first2;
+        }
```
Comment (strega-nil) [inline src/vcpkg/base/parse.cpp:189]:
```suggestion
        auto s_first = m_it.pointer_to_current();
        auto s_last = m_it.end().pointer_to_current();
        
        if (Strings::starts_with(StringView{s_first, s_last}, keyword_content))
        {
            auto end_of_keyword = s_first + keyword_content.size();
            if (end_of_keyword == s_last || is_whitespace(*end_of_keyword))
            {
                m_it = Utf8Decoder{end_of_keyword, s_last};
                m_column += static_cast<int>(keyword_content.size());
                return true;
            }
        }
        return false;
```

File: src/vcpkg/commands.ci.cpp [hunk 10]
```diff
@@ -487,12 +530,33 @@ namespace vcpkg::Commands::CI
...
+            const auto ci_baseline_file_contents =
+                paths.get_filesystem().read_contents(ci_baseline_file_name, VCPKG_LINE_INFO);
+            ParseMessages ci_parse_messages;
+            const auto lines = parse_ci_baseline(ci_baseline_file_contents, ci_baseline_file_name, ci_parse_messages);
+            ci_parse_messages.exit_if_errors_or_warnings(ci_baseline_file_name);
+            expected_failures = parse_and_apply_ci_baseline(lines, exclusions_map);
+        }
+        else
+        {
+            Checks::check_exit(VCPKG_LINE_INFO,
```
Comment (strega-nil) [inline src/vcpkg/commands.ci.cpp:552]: should this be localized?

**Review by strega-nil — CHANGES_REQUESTED** [review]
The one I want changed is `try_match_keyword`.

File: src/vcpkg/base/parse.cpp (hunk 9 above)
Comment (BillyONeal) [reply src/vcpkg/base/parse.cpp:189]: I think this is technically a change in behavior because is_whitespace is operating in the encoding unit domain rather than the code point domain, so whitespace like `U+2029 PARAGRAPH SEPARATOR` can't be interpreted as whitespace.

I don't necessarily care about that but want to make sure you saw and considered this before applying.

Comment (BillyONeal) [reply src/vcpkg/base/parse.cpp:189]: I did not apply this change because `m_it.end().pointer_to_current()` does not compile

File: src/vcpkg/commands.ci.cpp (hunk 10 above)
Comment (BillyONeal) [reply src/vcpkg/commands.ci.cpp]: Done.

File: include/vcpkg/base/parse.h (hunk 7 above)
Comment (BillyONeal) [reply include/vcpkg/base/parse.h]: So reverted.

File: src/vcpkg/base/parse.cpp (hunk 9 above)
Comment (strega-nil) [reply src/vcpkg/base/parse.cpp:190]: hmm, point. Can you do

```cxx
auto end_of_keyword = Utf8Decoder{s_first + keyword_content.size(), s_last};
if (end_of_keyword.is_eof() || is_whitespace(*end_of_keyword))
{
    m_it = end_of_keyword;
    ...
}
```

?

Comment (BillyONeal) [reply src/vcpkg/base/parse.cpp:190]: That's getting into territory where I'm 'scared' about the invariants of Utf8Decoder; the original solution carefully avoids crossing from decoded world back into encoding unit world. Let me try adding some comments and see if you like that...

Comment (BillyONeal) [reply src/vcpkg/base/parse.cpp]: Please let me know what you like of the solution I just pushed in https://github.com/microsoft/vcpkg-tool/pull/345/commits/19ae76dad6fecfb458764416a8346ddde836e6b6

Comment (strega-nil) [reply src/vcpkg/base/parse.cpp]: yeah, wfm

File: src/vcpkg-test/ci-baseline.cpp [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+static void check_error(const std::string& input, const std::string& expected_error)
+{
+    ParseMessages m;
+    auto actual = parse_ci_baseline(input, "test", m);
+    CHECK(actual.empty());
+    CHECK(m.warnings.empty());
+    CHECK(m.error->format() == expected_error);
+}
+
+TEST_CASE ("Parse Errors", "[ci-baseline]")
```
Comment (ras0219-msft) [inline src/vcpkg-test/ci-baseline.cpp:207]: Two cases I don't see here:
```
example # hello
:x64-windows # world
=
pass
```
and
```
example:x64-windows=fail example:x64-windows=fail example:x64-windows=fail
```

Comment (BillyONeal) [reply src/vcpkg-test/ci-baseline.cpp:207]: Added

**Comment by BillyONeal** [comment]
Thanks @autoantwort :D
